File: Python/Algorithms/1512.py
class Solution:
    def numIdenticalPairs(self, nums: List[int]) -> int:
        # sort initially.
        nums.sort()
        # 1 + 2 + 3 + ... + n
        sn = lambda n: (n * (n + 1)) // 2
        num_of_pairs = num = 0
        for i in range(1, len(nums)):
            # found a pair
            if nums[i] == nums[i - 1]:
                num += 1
            # bumped into next group.
            else:
                num_of_pairs += sn(num)
                num = 0
        # add left-overs from final group.
        num_of_pairs += sn(num)
        return num_of_pairs

File: Rust/Algorithms/1512.rs
impl Solution {
    pub fn num_identical_pairs(nums: Vec<i32>) -> i32 {
        let mut mnums = nums.to_owned();
        // need mut handle to sort.
        mnums.sort();
        
        // 1 + 2 + 3 + ... + n
        let sn = |n| (n * (n + 1)) / 2;
        let (mut num_pairs, mut num) = (0, 0);
        
        // Go through vector.
        for i in 1..mnums.len(){
            // find len of group.
            if (mnums[i] == mnums[i-1]){
                num += 1;
            }
            // add sum and start again.
            else {
                num_pairs += sn(num);
                num = 0;
            }
        }
        // add leftovers.
        num_pairs += sn(num);
        return num_pairs;
    }
}
